modules/linear_boltzmann_solvers/discrete_ordinates_problem/sweep_chunks/cbc_sweep_chunk.cc

```diff
@@ -205,116 +259,203 @@ CBCSweepChunk::Sweep(AngleSet& angle_set)
       const bool is_boundary_face = not face.has_neighbor;
       const bool is_reflecting_boundary_face =
         (is_boundary_face and angle_set.GetBoundaries()[face.neighbor_id]->IsReflecting());
-      const auto& IntF_shapeI = IntS_shapeI_[f];
 
-      const int locality = cell_transport_view_->FaceLocality(f);
-      const size_t num_face_nodes = cell_mapping_->GetNumFaceNodes(f);
-      auto& face_nodal_mapping = fluds_->GetCommonData().GetFaceNodalMapping(cell_local_id_, f);
-      std::vector<double>* psi_dnwnd_data = nullptr;
+      /// Surface integral of shape functions for face f
+      const auto& IntFi_shapeI_vec = IntS_shapeI_[f];
+
+      /// Prepare buffer for remote downwind data
+      std::vector<double>* psi_nonlocal_dnwnd_data_block_ptr = nullptr;
+
       if (not is_boundary_face and not is_local_face)
-      {
+      { /// If remote face
+        const int locality = cell_transport_view_->FaceLocality(f);
+        auto& face_nodal_mapping = fluds_->GetCommonData().GetFaceNodalMapping(cell_local_id_, f);
         auto& async_comm = *angle_set.GetCommunicator();
-        size_t data_size = num_face_nodes * group_angle_stride_;
-        psi_dnwnd_data = &async_comm.InitGetDownwindMessageData(locality,
-                                                                face.neighbor_id,
-                                                                face_nodal_mapping.associated_face_,
-                                                                angle_set.GetID(),
-                                                                data_size);
+
+        /// Remote downwind buffer sizing (for multi-AngleSet-angle packet)
+        /// Size of buffer for ALL angles in this AngleSet, ALL groups,
+        /// for ALL nodes on this face
+        /// remote_node_stride_ = num_angles_in_set_remote_ * gs_size_
+        const size_t data_size_for_msg = num_face_nodes * remote_node_stride_;
+
+        psi_nonlocal_dnwnd_data_block_ptr = &async_comm.InitGetDownwindMessageData(
+          locality,
+          face.neighbor_id,
+          face_nodal_mapping.associated_face_,
+          angle_set.GetID(),
+          data_size_for_msg); /// Request buffer for multi-angle packet
       }
 
+      /// Loop over nodes ON THIS FACE of current cell
       for (int fi = 0; fi < num_face_nodes; ++fi)
       {
+        /// Node index in current cell
         const int i = cell_mapping_->MapFaceNode(f, fi);
 
+        /// Tally outflow for particle balance (for ALL boundary faces)
         if (is_boundary_face)
         {
+          /// Integral of shape func for node i on face f
+          const double IntFi_shapeI_val = IntFi_shapeI_vec(i);
           for (int gsg = 0; gsg < gs_size_; ++gsg)
             cell_transport_view_->AddOutflow(
-              f, gs_gi_ + gsg, wt * face_mu_values[f] * b[gsg](i) * IntF_shapeI(i));
+              f, gs_gi_ + gsg, wt * face_mu_values[f] * b[gsg](i) * IntFi_shapeI_val);
```

Comment:
going from `IntF_shapeI(i)` to `IntFi_shapeI_vec(i)` seems like a bug fix. Correct? 
@wdhawkins , what do you think?